### PR TITLE
Clean up output directory before copying over new files

### LINKED ISSUE #13878: Assertion '../../third_party/dart/runtime/vm/kernel_binary.cc: 20: error: Invalid magic identifier' during hot reload in default mode after running with --preview-dart-2
## Steps to Reproduce

1. Build local engine in 'unoptimized' mode:

```
flutter/tools/gn --goma --android --unoptimized --runtime-mode=debug && ninja -j1000 -C out/android_debug_unopt
```

2. Run flutter_gallery app with --preview-dart-2, wait for app to appear, press 'q'.

```
flutter clean
flutter run --preview-dart-2
```

3. Run flutter_gallery app in debug mode with local engine, without --preview-dart-2 and *without* cleaning the build directory; wait for app to appear, press 'r' for hot reload:

```
flutter run --local-engine android_debug_unopt
```

Output:
```
Initializing hot reload...                                
E/DartVM  ( 6160): ../../third_party/dart/runtime/vm/kernel_binary.cc: 20: error: Invalid magic identifier
E/DartVM  ( 6160): Dumping native stack trace for thread 181f
E/DartVM  ( 6160):   [0xa20f7629] Unknown symbol
E/DartVM  ( 6160):   [0xa20f7629] Unknown symbol
E/DartVM  ( 6160):   [0xa2390f41] Unknown symbol
E/DartVM  ( 6160):   [0xa20109db] Unknown symbol
E/DartVM  ( 6160):   [0xa2010c31] Unknown symbol
E/DartVM  ( 6160):   [0xa2285c57] Unknown symbol
E/DartVM  ( 6160):   [0xa1e8eae9] Unknown symbol
E/DartVM  ( 6160):   [0xa1e8e025] Unknown symbol
E/DartVM  ( 6160):   [0xa1e8b887] Unknown symbol
E/DartVM  ( 6160):   [0xa2005599] Unknown symbol
E/DartVM  ( 6160):   [0xa1ffeac7] Unknown symbol
E/DartVM  ( 6160):   [0xa215268d] Unknown symbol
E/DartVM  ( 6160):   [0xa214b58f] Unknown symbol
E/DartVM  ( 6160):   [0xa214bc49] Unknown symbol
E/DartVM  ( 6160):   [0xa1ffb7d5] Unknown symbol
E/DartVM  ( 6160):   [0xa2019927] Unknown symbol
E/DartVM  ( 6160):   [0xa20199cd] Unknown symbol
E/DartVM  ( 6160):   [0xa226901f] Unknown symbol
E/DartVM  ( 6160):   [0xa1e8a201] Unknown symbol
E/DartVM  ( 6160):   [0xa1e8a863] Unknown symbol
E/DartVM  ( 6160):   [0xa1e8a805] Unknown symbol
E/DartVM  ( 6160):   [0xa148ccdf] Unknown symbol
E/DartVM  ( 6160):   [0xa148c5f7] Unknown symbol
E/DartVM  ( 6160):   [0xa148c13b] Unknown symbol
E/DartVM  ( 6160):   [0xa1495685] Unknown symbol
E/DartVM  ( 6160):   [0xa14956f1] Unknown symbol
E/DartVM  ( 6160):   [0xa14956b9] Unknown symbol
E/DartVM  ( 6160): -- End of DumpStackTrace
F/libc    ( 6160): Fatal signal 6 (SIGABRT), code -6 in tid 6175 (ui_thread)
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'motorola/athene/athene:6.0.1/MPJ24.139-64/65:user/release-keys'
Revision: 'p2a0'
ABI: 'arm'
pid: 6160, tid: 6175, name: ui_thread  >>> io.flutter.examples.gallery <<<
signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
    r0 00000000  r1 0000181f  r2 00000006  r3 af7fc978
    r4 af7fc980  r5 af7fc930  r6 0000000b  r7 0000010c
    r8 a249ea2c  r9 b81879f8  sl b80f9d78  fp b8187978
    ip 00000006  sp af7f9d78  lr b6d471cd  pc b6d495c8  cpsr 400f0010
backtrace:
    #00 pc 000445c8  /system/lib/libc.so (tgkill+12)
    #01 pc 000421c9  /system/lib/libc.so (pthread_kill+32)
    #02 pc 0001bb1f  /system/lib/libc.so (raise+10)
    #03 pc 00018cc1  /system/lib/libc.so (__libc_android_abort+34)
    #04 pc 00016880  /system/lib/libc.so (abort+4)
    #05 pc 003a081f  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #06 pc 00fd1f43  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #07 pc 00c519d7  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #08 pc 00c51c2d  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #09 pc 00ec6c53  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #10 pc 00acfae5  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #11 pc 00acf021  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #12 pc 00acc883  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #13 pc 00c46597  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #14 pc 00c3fac3  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #15 pc 00d93689  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #16 pc 00d8c58d  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #17 pc 00d8cc45  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #18 pc 00c3c7d1  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #19 pc 00c5a925  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #20 pc 00c5a9c9  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #21 pc 00eaa01b  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #22 pc 00acb1fd  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #23 pc 00acb85f  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #24 pc 00acb801  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #25 pc 000cdcdd  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #26 pc 000cd5f3  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #27 pc 000cd137  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #28 pc 000d6681  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #29 pc 000d66ed  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #30 pc 000d66b5  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #31 pc 00012ebf  /system/lib/libutils.so (_ZN7android6Looper9pollInnerEi+530)
    #32 pc 00012f8f  /system/lib/libutils.so (_ZN7android6Looper8pollOnceEiPiS1_PPv+130)
    #33 pc 0000885d  /system/lib/libandroid.so (ALooper_pollOnce+64)
    #34 pc 000d655b  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #35 pc 000cda3d  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #36 pc 000cc3ad  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #37 pc 000d16f7  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #38 pc 000d15b3  /data/app/io.flutter.examples.gallery-1/lib/arm/libflutter.so
    #39 pc 00041acb  /system/lib/libc.so (_ZL15__pthread_startPv+30)
    #40 pc 00019355  /system/lib/libc.so (__start_thread+6)
Tombstone written to: /data/tombstones/tombstone_09
Lost connection to device.
```

## Flutter Doctor

[✓] Flutter (on Linux, locale en_US.UTF-8, channel unknown)
    • Flutter at /usr/local/google/home/alexmarkov/work/flutter
    • Framework revision 1eb6a95fcb (3 hours ago), 2018-01-03 23:57:38 +0800
    • Engine revision 37789650cb
    • Tools Dart version 1.25.0-dev.11.0
    • Engine Dart version 2.0.0-edge.e122f4825726035e4579471ef83848b1aa796444

[✓] Android toolchain - develop for Android devices (Android SDK 22.0.1)
    • Android SDK at /usr/local/google/home/alexmarkov/work/engine/src/third_party/android_tools/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-22, build-tools 22.0.1
    • ANDROID_HOME = /usr/local/google/home/alexmarkov/work/engine/src/third_party/android_tools/sdk
    • Java binary at: /opt/android-studio-3-preview/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-884-b01)

[✓] Android Studio (version 3.0)
    • Android Studio at /opt/android-studio-3-preview
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-884-b01)

[✓] Android Studio (version 2.3)
    • Android Studio at /opt/android-studio
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Connected devices
    • Moto G 4 • ZY2245H6M5 • android-arm • Android 6.0.1 (API 23)

### PR DESCRIPTION
**Pull request by aam**
Clean up of whitespaces aside, this deletes output directory before copying new build artifacts over.

This complements https://github.com/flutter/flutter/pull/14081 that does the same thing on ios, which is needed to prevent incompatible(`--preview-dart-2` vs non-`preview-dart-2`) build artifacts being included into the built app.

This fixes https://github.com/flutter/flutter/issues/13878.

### CODE REVIEW

**Review by cbracken — APPROVED** [review]

**Comment by cbracken** [comment]
I'm not a Gradle expert, but if this does what it purports to do, LGTM. 

/cc @mravn-google and @szakarias who are the Gradle experts.

File: packages/flutter_tools/gradle/flutter.gradle [hunk 1]
```diff
@@ -310,17 +310,22 @@ class FlutterPlugin implements Plugin<Project> {
...
                 preferSharedLibrary preferSharedLibraryValue
                 sourceDir project.file(project.flutter.source)
                 intermediateDir project.file("${project.buildDir}/${AndroidProject.FD_INTERMEDIATES}/flutter/${variant.name}")
                 extraFrontEndOptions extraFrontEndOptionsValue
                 extraGenSnapshotOptions extraGenSnapshotOptionsValue
             }
 
+            Task deleteOutputDir = project.tasks.create(name: "deleteOutputDirectory${variant.name.capitalize()}", type:Delete) {
+                delete variant.mergeAssets.outputDir
+            }
```
Comment (mravn-google) [inline packages/flutter_tools/gradle/flutter.gradle]: I believe there is a `cleanXxx` task for every output-producing task `xxx`, so this should not be necessary. See below.

File: packages/flutter_tools/gradle/flutter.gradle [hunk 2]
```diff
@@ -310,17 +310,22 @@ class FlutterPlugin implements Plugin<Project> {
...
+
             Task copyFlxTask = project.tasks.create(name: "copyFlutterAssets${variant.name.capitalize()}", type: Copy) {
                 dependsOn flutterTask
                 dependsOn variant.mergeAssets
+                dependsOn deleteOutputDir
```
Comment (mravn-google) [inline packages/flutter_tools/gradle/flutter.gradle]: You should be able to replace this line with
```gradle
            dependsOn "clean${variant.mergeAssets.name.capitalize()}"
```

**Review by mravn-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/gradle/flutter.gradle (hunk 1 above)
Comment (aam) [reply packages/flutter_tools/gradle/flutter.gradle]: Cool, good to know.

File: packages/flutter_tools/gradle/flutter.gradle (hunk 2 above)
Comment (aam) [reply packages/flutter_tools/gradle/flutter.gradle]: Done

**Pull request merged**